jupyterhub_traefik_proxy/traefik_utils.py

```diff
@@ -1,107 +1,139 @@
-            auth_password=password,
-        )
-        backends_data = json.loads(resp_backends.body)
-        frontends_data = json.loads(resp_frontends.body)
-
-        if resp_backends.code == 200 and resp_frontends.code == 200:
-            ready = (
-                expected_backend in backends_data
-                and expected_frontend in frontends_data
-            )
-    except Exception as e:
-        backends_rc, frontends_rc = e.response.code
-        ready = False
-    finally:
-        return ready
-
-
-async def check_traefik_static_conf_ready(username, password, api_url):
-    """ Check if traefik loaded its static configuration from the
-    etcd cluster """
-    try:
-        resp = await AsyncHTTPClient().fetch(
-            api_url + "/api/providers/etcdv3",
-            auth_username=username,
-            auth_password=password,
-        )
-        rc = resp.code
-    except ConnectionRefusedError:
-        rc = None
-    except Exception as e:
-        rc = e.response.code
-    finally:
-        return rc == 200
-
-
-def launch_traefik_with_toml():
-    config_file_path = abspath(join(dirname(__file__), "traefik.toml"))
-    traefik = Popen(["traefik", "-c", config_file_path], stdout=None)
-    return traefik
-
+from contextlib import contextmanager
+from collections import namedtuple
+
+
+def generate_rule(routespec):
+    if routespec.startswith("/"):
+        # Path-based route, e.g. /proxy/path/
+        rule = "PathPrefix:" + routespec
+    else:
+        # Host-based routing, e.g. host.tld/proxy/path/
+        host, path_prefix = routespec.split("/", 1)
+        path_prefix = "/" + path_prefix
+        rule = "Host:" + host + ";PathPrefix:" + path_prefix
+    return rule
+
+
+def generate_alias(url, server_type=""):
+    return server_type + re.sub("[.:/]", "_", (urlparse(url).netloc))
+
+
+def generate_backend_entry(
+    proxy, backend_alias, separator="/", url=False, weight=False
+):
+    backend_entry = ""
+    if separator is "/":
+        backend_entry = proxy.etcd_traefik_prefix
+    backend_entry += (
+        "backends"
+        + separator
+        + backend_alias
+        + separator
+        + "servers"
+        + separator
+        + "server1"
+    )
+    if url is True:
+        backend_entry += separator + "url"
+    elif weight is True:
+        backend_entry += separator + "weight"
 
-def launch_traefik_with_etcd():
-    traefik = Popen(["traefik", "--etcd", "--etcd.useapiv3=true"], stdout=None)
-    return traefik
+    return backend_entry
 
 
-def generate_traefik_toml():
-    pass  # Not implemented
+def generate_frontend_backend_entry(proxy, frontend_alias):
+    return proxy.etcd_traefik_prefix + "frontends/" + frontend_alias + "/backend"
 
 
-def create_backend_alias_from_url(url):
-    target = urlparse(url)
-    return "jupyterhub_backend_" + target.netloc
+def generate_frontend_rule_entry(proxy, frontend_alias, separator="/"):
+    frontend_rule_entry = (
```

Comment:
This can be written as:

```python
frontend_rule_entry = separator.join(["frontends", frontend_alias, "routes", "test"])
```

to avoid writing separator over and over again.